### ISSUE TITLE
Local Minimal Development

### ISSUE THREAD

**Issue by hongxuzhou** [hongxuzhou/illusion-of-state#12]
**The code should run without any bugs in my local env**

* Transformer skeleton: nanoGPT style, supporting configuring numbers of layers/dimension/head; verify the shape output by the forward pass is correct. 
* Mamba Skeleton: mamba-minimal?? [The important thing is to find a minimal architecture that supports both CUDA and MPS], the interface that aligns the parameters with the transformer. Also, verify the shape
* Train loop: support early stopping, saving checkpoints, and learning rates. Run the models on Task A in the local env till convergence (no need to have a good accuracy, but need to witness loss lowering down)
* Evaluate pipeline: `evaluate.py` outputs sequence-level accuracy and per-length accuracy. Verify that the output format is correct on random models.
* Probing skeleton: the hook mechanism in `extract_states.py` can accurately extract the hidden states of the correct shapes. No need to train the probe, but need to check the dimension. 
* Local end-to-end test: Train transformer on task A small data till overfitting, acuracy >95, proving this this model's compacity instead of coding issue.
